Add rudimentary implementation of imperative slot API

rniwa opened the pull request:
#### b3a5c931e34b75897600609a07df70e3dc0924cc
<pre>
Add rudimentary implementation of imperative slot API
<a href="https://bugs.webkit.org/show_bug.cgi?id=243772">https://bugs.webkit.org/show_bug.cgi?id=243772</a>

Reviewed by Antti Koivisto.

Add a very basic implementation of imperative slot API.

NodeRareData now contains a back reference to the slot element to which the node is manually assigned.
Each slot element remembers which nodes were last manually assigned.

These two states are updated by ManualSlotAssignment.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-slotchange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-fallback-clear-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:

* Source/WebCore/dom/Node.cpp:
(WebCore::stringForRareDataUseType):
(WebCore::Node::manuallyAssignedSlot const): Added.
(WebCore::Node::setManuallyAssignedSlot): Added.
* Source/WebCore/dom/Node.h:

* Source/WebCore/dom/NodeRareData.cpp:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeRareData::manuallyAssignedSlot): Added.
(WebCore::NodeRareData::setManuallyAssignedSlot): Added.
(WebCore::NodeRareData::useTypes const):

* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::addSlotElementByName): Construct ManualSlotAssignment as needed.
(WebCore::ShadowRoot::slotManualAssignmentDidChange):
* Source/WebCore/dom/ShadowRoot.h:

* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::slotManualAssignmentDidChange): Added. No-op.
(WebCore::ManualSlotAssignment::findAssignedSlot): Ditto.
(WebCore::effectiveAssignedNodes): Ditto.
(WebCore::ManualSlotAssignment::assignedNodesForSlot): Ditto.
(WebCore::ManualSlotAssignment::renameSlotElement): Added. No-op.
(WebCore::ManualSlotAssignment::addSlotElementByName): Ditto.
(WebCore::ManualSlotAssignment::removeSlotElementByName): Ditto.
(WebCore::ManualSlotAssignment::slotManualAssignmentDidChange): Added. This implements the main logic
for updating manually assigned nodes. First, we clear slot-back references in the previous assignment,
and destroy any renderer created for the previous assignment. Second, we go through the current assignment
and find any slot in the same shadow tree from which we&apos;re taking nodes away. Finally, we visit each
slot which got affected by this assignment in the tree order and enqueue slotchange events as needed.
(WebCore::ManualSlotAssignment::slotFallbackDidChange): Added.
(WebCore::ManualSlotAssignment::hostChildElementDidChange): Ditto.
(WebCore::ManualSlotAssignment::hostChildElementDidChangeSlotAttribute): Ditto.
(WebCore::ManualSlotAssignment::willRemoveAssignedNode): Ditto.
(WebCore::ManualSlotAssignment::didRemoveAllChildrenOfShadowHost): Ditto.
(WebCore::ManualSlotAssignment::didMutateTextNodesOfShadowHost): Ditto.
* Source/WebCore/dom/SlotAssignment.h:
(WebCore::ManualSlotAssignment): Added.
(WebCore::ManualSlotAssignment::Slot): Added.

* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assign): Added.
(WebCore::HTMLSlotElement::removeManuallyAssignedNode): Added.
* Source/WebCore/html/HTMLSlotElement.h:

* Source/WebCore/html/HTMLSlotElement.idl:

Canonical link: <a href="https://commits.webkit.org/253320@main">https://commits.webkit.org/253320@main</a>
</pre>

anttijk commented on Source/WebCore/dom/SlotAssignment.h:
```diff
@@ -135,6 +137,35 @@ class NamedSlotAssignment : public SlotAssignment {
+        uint32_t cachedVersion { 0 };
+    };
+    WeakHashMap<HTMLSlotElement, Slot> m_slots;
+    uint32_t m_slottableVersion { 0 };
```
These are probably not going to overrun but there is probably no reason not to just use uin64_t either.

rniwa replied:
Ok, I'd change it to uint64_t.

anttijk commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -115,10 +115,10 @@ static const char* stringForRareDataUseType(NodeRareData::UseType useType)
         return "NodeList";
     case NodeRareData::UseType::MutationObserver:
         return "MutationObserver";
+    case NodeRareData::UseType::ManuallyAssignedSlot:
```
Is Is "manually assigned" a spec term? If not then how about "ImperativelyAssigned" or something (here and everywhere else)? I don't think "manual" is a good way to describe something done via API.

rniwa replied:
Yes, manually assigned is a spec term:
https://html.spec.whatwg.org/multipage/scripting.html#manually-assigned-nodes
I'd add a link to this spec in the header.

rniwa committed:
Add rudimentary implementation of imperative slot API
https://bugs.webkit.org/show_bug.cgi?id=243772

Reviewed by Antti Koivisto.

Add a very basic implementation of imperative slot API.

NodeRareData now contains a back reference to the slot element to which the node is manually assigned.
Each slot element remembers which nodes were last manually assigned.

These two states are updated by ManualSlotAssignment.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-api-slotchange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/imperative-slot-fallback-clear-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:

* Source/WebCore/dom/Node.cpp:
(WebCore::stringForRareDataUseType):
(WebCore::Node::manuallyAssignedSlot const): Added.
(WebCore::Node::setManuallyAssignedSlot): Added.
* Source/WebCore/dom/Node.h:

* Source/WebCore/dom/NodeRareData.cpp:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeRareData::manuallyAssignedSlot): Added.
(WebCore::NodeRareData::setManuallyAssignedSlot): Added.
(WebCore::NodeRareData::useTypes const):

* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::addSlotElementByName): Construct ManualSlotAssignment as needed.
(WebCore::ShadowRoot::slotManualAssignmentDidChange):
* Source/WebCore/dom/ShadowRoot.h:

* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::slotManualAssignmentDidChange): Added. No-op.
(WebCore::ManualSlotAssignment::findAssignedSlot): Ditto.
(WebCore::effectiveAssignedNodes): Ditto.
(WebCore::ManualSlotAssignment::assignedNodesForSlot): Ditto.
(WebCore::ManualSlotAssignment::renameSlotElement): Added. No-op.
(WebCore::ManualSlotAssignment::addSlotElementByName): Ditto.
(WebCore::ManualSlotAssignment::removeSlotElementByName): Ditto.
(WebCore::ManualSlotAssignment::slotManualAssignmentDidChange): Added. This implements the main logic
for updating manually assigned nodes. First, we clear slot-back references in the previous assignment,
and destroy any renderer created for the previous assignment. Second, we go through the current assignment
and find any slot in the same shadow tree from which we're taking nodes away. Finally, we visit each
slot which got affected by this assignment in the tree order and enqueue slotchange events as needed.
(WebCore::ManualSlotAssignment::slotFallbackDidChange): Added.
(WebCore::ManualSlotAssignment::hostChildElementDidChange): Ditto.
(WebCore::ManualSlotAssignment::hostChildElementDidChangeSlotAttribute): Ditto.
(WebCore::ManualSlotAssignment::willRemoveAssignedNode): Ditto.
(WebCore::ManualSlotAssignment::didRemoveAllChildrenOfShadowHost): Ditto.
(WebCore::ManualSlotAssignment::didMutateTextNodesOfShadowHost): Ditto.
* Source/WebCore/dom/SlotAssignment.h:
(WebCore::ManualSlotAssignment): Added.
(WebCore::ManualSlotAssignment::Slot): Added.

* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assign): Added.
(WebCore::HTMLSlotElement::removeManuallyAssignedNode): Added.
* Source/WebCore/html/HTMLSlotElement.h:

* Source/WebCore/html/HTMLSlotElement.idl:

Canonical link: https://commits.webkit.org/253320@main
